### ISSUE TITLE
If Intellicode Whole Line Completions are displayed, the suggestions widget does not open

### ISSUE THREAD

**Issue by juihanamshet1** [microsoft/vscode#131940]
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.60.0-insider (user setup)
- OS Version: Windows_NT x64 10.0.19043
- List of extensions that you have installed: Visual Studio IntelliCode Completions

Steps to Reproduce:

1. Open the attached .txt file, convert it to a typescript .ts file 
[main.txt](https://github.com/microsoft/vscode/files/7079987/main.txt)

2. In line 5, type `s=` and you should see a gray text IntelliCode completion. 
3. When you type what the completion shows, and get to typing a dot, ".".
4. When you type a ".", it is expected you see the suggestions widget. But the suggestions widget does not show up in the presence of gray text completion. 
5. GIF - ![completionsDrivenBug2](https://user-images.githubusercontent.com/86270736/131415373-5761e82f-51ce-42c1-bd85-a501ef9a54bf.gif)

**Comment by juihanamshet1** [comment]
@hediet

**Comment by hediet** [comment]
This is actually by design (the ghost text should not disappear when typing the text as suggested).
However, I want to change this so that autocomplete shows, but the item from the ghost text is preselected.

**Comment by juihanamshet1** [comment]
"the item from the ghost text is preselected", can you please clarify this? In the above scenario, since ".socket" is in the whole line completion, would "socket" be selected in the suggestions widget?

**Comment by hediet** [comment]
>  In the above scenario, since ".socket" is in the whole line completion, would "socket" be selected in the suggestions widget?

Yes, when the autocomplete opens automatically while the cursor is in ghost text, the preselected item should agree with the ghost text.
Otherwise the ghost text will just disappear when the autocomplete opens.

**Comment by juihanamshet1** [comment]
That sounds great! Do you have an estimate as to when this would be implemented?

**Comment by hediet** [comment]
> Do you have an estimate as to when this would be implemented?

This milestone hopefully (i.e. early October).

**Comment by Krzysztof-Cieslak** [comment]
There's no guarantee that result returned by the Copilot is in the list of the completions for given location

**Comment by hediet** [comment]
> There's no guarantee that result returned by the Copilot is in the list of the completions for given location

You mean that a non-empty prefix of a copilot suggestion appears in the suggest widget?
Do you have an example where that could be problematic?

**Comment by hediet** [comment]
@Krzysztof-Cieslak @juihanamshet1 the feature will be available in tomorrows insiders build of VS Code.
I would love to hear your feedback!

**Comment by lramos15** [comment]
I see there's a test plan item for this so marking as on-tesplan instead of verification-needed

**Comment by orph** [comment]
@hediet this is the first I've heard of this, sorry I missed it. Can you please verify that the InlineCompletionItem itself will be listed in the popup? Since we don't currently register a regular completion item handler, only the inline version, will this mean a valid inline completion will just stop displaying in order to show the completion popup?

/cc @Krzysztof-Cieslak @isidorn

**Comment by hediet** [comment]
> Since we don't currently register a regular completion item handler, only the inline version, will this mean a valid inline completion will just stop displaying in order to show the completion popup?

A "valid" inline completion item will only be not shown with this new behavior if no proper prefix of it is suggested by the autocomplete.
If no prefix of the inline completion is a valid auto-complete item, I think the inline completion is very likely to cause compile errors.

**Comment by hsenag** [comment]
> If no prefix of the inline completion is a valid auto-complete item, I think the inline completion is very likely to cause compile errors.

@hediet I guess that might depend on how good the auto-complete info is for a particular language. Is this behaviour going to be enabled for all languages?

Also is there any event when an inline completion item gets rejected by this logic?

**Comment by hediet** [comment]
> Is this behaviour going to be enabled for all languages?

Yes.

> Also is there any event when an inline completion item gets rejected by this logic?

No - they are not rejected, simply just not shown.

> I guess that might depend on how good the auto-complete info is for a particular language. 

Do you have an example where this is not good?
I've been using this feature for 20 days now and it seems to work nicely with TypeScript.

**Comment by hsenag** [comment]
> No - they are not rejected, simply just not shown.

I guess the terminology is a matter of perspective :-) From the point of view of our extension which is providing inline completion items, it'd be really useful to know if VSCode decided not to show them and why.

> Do you have an example where this is not good?

I just had a quick play with Haskell and it offers auto-complete suggestions when I start writing a new function. I don't know what the general expectations around auto-complete are and this is arguably a bug in the language server or whatever is providing those suggestions, but it's the behaviour we have in practice today.

**Comment by hediet** [comment]
> I just had a quick play with Haskell and it offers auto-complete suggestions when I start writing a new function.

Can you provide a screen recording/screenshot that shows why the auto-complete interfers with copilot in this case?
Does copilot suggest something that has no prefix which is suggested by auto-complete?
I would like to understand this problem better.

**Comment by hsenag** [comment]
Here's a simple example:

https://user-images.githubusercontent.com/451141/135254469-bb0c3ac9-f23c-4c33-a695-91166a213aa8.mp4

I setup the prompt so that it's obvious to the model that `myFunc2` needs to be implemented.

Then I start typing `m` but auto-complete pops up with a bunch of (wrong) suggestions. Then I hit esc to cancel auto-complete and I see the Copilot suggestion, which is correct.

**Comment by Krzysztof-Cieslak** [comment]
So let's look at this - here Copilot is making a suggestion based on the existing patterns in users code. `asd`, `bsd` and `csd` functions currently don't exist in accessible scope so they won't be present in the autocomplete list.

However, maybe the user just plan to implement them in the future? Or maybe they're just coming from different packages/files and all that's needed is an import statement? Copilot suggestions are not driven by semantic information, they potentially can produce not valid code... but it doesn't mean they are wrong.

https://user-images.githubusercontent.com/5427083/135254275-1437b4d6-6c8b-4f47-92c5-912694f1b463.mov

**Comment by hediet** [comment]
Can you set `editor.inlineSuggest.allowQuickSuggestions` to `false` and check if that is the desired behavior? Also pinging @juihanamshet1

That way, trigger characters still cause the autocomplete to show up while it does not show up when simply typing letters:

![recording](https://user-images.githubusercontent.com/2931520/135259347-d4d17d04-7282-4237-bbb7-dfcda2f8857c.gif)

I think this would address all issues & feature requests.

We can set `allowQuickSuggestions` to `false`.

**Comment by hsenag** [comment]
Did you mean `editor.quickSuggestions` (not `editor.inlineSuggest.allowQuickSuggestions`)? If so that works around this problem as auto complete doesn't appear, but it also disables it in scenarios where it does provide useful suggestions. Of course users will get it when there's a trigger character or can hit Ctrl-Space.

My feeling is that the root cause here is that both auto-complete and inline suggestion providers like Copilot are never right 100% of the time.

Where the two are not in conflict, it's a great user experience now because they still see auto-complete and the inline completion and can choose which direction to explore (accept the full inline completion, or reject it and pick a different auto-complete item).

It's where they disagree that there's a problem and it's hard for VS Code to know which one is right, particularly given the examples of currently invalid code that @Krzysztof-Cieslak mentioned above. Is there some way to present both to the user even when they do disagree? e.g. an empty item in auto-complete?

**Comment by hsenag** [comment]
I tried `editor.inlineSuggest.allowQuickSuggestions` and it does help. I guess that when a trigger character like `.` is pressed, the behaviour still relies on the auto-complete suggestions actually being complete?

**Comment by hediet** [comment]
> I guess that when a trigger character like . is pressed, the behaviour still relies on the auto-complete suggestions actually being complete?

Yes. However, we can iterate on that.

I'm also very interested in collecting (counter-) examples where the auto-complete suggestions are not complete (and missing the suggestion of copilot, or a prefix of it) when the user presses the trigger character.

**Comment by hediet** [comment]
Maybe we can explore only showing the suggest widget if there is an item that is a prefix of the inline suggestions.

**Comment by ayim** [comment]
@hediet here's what I've noticed so far...

**Default Behaviour**

Condition:

- When editing in languages with a language server
- AND without any grey-text extension installed

Behaviour:

The IntelliSense menu shows up as the user is typing (not just on trigger characters) **

**What we're trying to build:**

**An option to allow opting out from the base VS Code IntelliSense behaviour while _greytext_ is shown could be a solution that makes everyone happy.**

1) For experiences where there's less grey text shown at once (i.e. IntelliCode), I think it makes sense to have IntelliSense shown while typing and on trigger characters. There's three benefits here:

- In these scenarios, this interaction allows the user steer the suggestion as they type, along with 'jogging' their memory with helpful reminders/hints
- We like to think of IntelliCode as an extension or an augmentation of the existing IntelliSense experience; given this, we prefer to keep the editing experience as similar to the base editing experience as possible

2) For the Copilot experience (and others) where up to a whole method can be predicted at once, the Intellisense list showing up can be particularly disruptive. For cases like this, I'm curious what the Copilot team's desired experience looks like.

- When a grey text suggestion is shown, does the Intellisense list show on some trigger commands (i.e. Ctrl + Space) or none at all?
- When a grey text suggestion is shown, does the Intellisense list show on some trigger characters (i.e. .) or none at all?

**Comment by ayim** [comment]
> Early prototype: ![recording](https://user-images.githubusercontent.com/2931520/132365060-8fb7fc7b-50a7-4688-b638-4dc537eed038.gif)
> 
> @juihanamshet1 @Krzysztof-Cieslak @orph what do you think about this? The suggest widget would no longer hide, but the inline suggestion is preselected, so that the ghost text does not disappear.

Yes! We would love to have this behaviour; we've implemented something similar on the Visual Studio side.

**Comment by orph** [comment]
Ok so here is what I think could be an awesome strategy to integrate autocomplete popup and inline completions more seamlessly, while also keeping the features separate as per requested Copilot user feedback:

1. when typing line the inline completion is shown and updates as you type
2. when an autocomplete trigger character is typed, the extension host request for completions items is made in the background
3. if the current inline completion starts with any of the returned completions item text, the AI is on track, so no popup is shown
4. otherwise the current inline completion does not match any completion items, the AI is off track, so hide the inline completion and show the autocomplete popup list

What do you think @hediet?

**Comment by hediet** [comment]
> 3. if the current inline completion starts with any of the returned completion item text, the AI is on track, so no popup is shown

Why do you want to hide the popup here? VS Code pre-selects the item in the popup and renders a preview for both the item and the remaining inline suggestion.
The only difference is:
1) Pressing "tab" first accepts the autocomplete, a second tab accepts the remaining inline suggestion
2) The popup is visible
3) The user can use the arrow keys to select a different item in the popup

As the user selects a different item, the inline suggestion provider can make use of that and provide an updated inline suggestion.

These are my thoughts:
1) Makes it annoying to accept the inline suggestion, as you have to press tab twice.
    * However, it makes it convenient to accept the autocomplete when the inline suggestion is only partially correct. This is a compromise and I think there is no clear winner.
2) Adds visual noise when focusing on inline suggestions, but also shows valuable information when focusing on auto-complete. Again a compromise.
3) This is independant of inline-suggestions, as arrow keys don't affect them. I would argue this is a feature without drawbacks.

**Comment by Krzysztof-Cieslak** [comment]
> Why do you want to hide the popup here? VS Code pre-selects the item in the popup and renders a preview for both the item and the remaining inline suggestion.

I think we're mostly worried that having both on the screen at the same time is visually distracting. It's based on the feedback we get from users in the early Copilot days (back when we were using Text Decorations API for ghost text). 

That said back then the features were not in-sync - I'm not sure how much of the complaints were related to the fact there were 2 things on the screen and how much to the fact there were **different** suggestions displayed

**Comment by hediet** [comment]
>  I'm not sure how much of the complaints were related to the fact there were 2 things on the screen and how much to the fact there were different suggestions displayed

Didn't you also show the inline suggestion as autocomplete entry?

I totally understand the feedback that it is very confusing if inline suggestions and the autocomplete operate fully independent. But this feature is about making them work nicely together.

**Comment by Krzysztof-Cieslak** [comment]
Additionally, I want to point out that what @orph has proposed wouldn't fix the issue described by @mtaran - in this case, we want to show a suggestion that is not on the list of completions. If pop up would be shown here (as described in point 4. in @orph's algorithm) we would "lose" it... 

Maybe the solution is always showing autocomplete on trigger characters, and if our suggestion is not on the list some dummy entry should be added to the list of suggestions (and it would be selected by default)?

**Comment by hediet** [comment]
> Maybe the solution is always showing autocomplete on trigger characters, and if our suggestion is not on the list, some dummy entry should be added to the list of suggestions (and it would be selected by default)?

I like that idea. I think that trigger characters are usually followed by identifiers (e.g. ID_START ID_CONT* as defined by unicode). That identifier could also be extracted from the inline suggestion.

Alternatively, we could not show the autocomplete if copilots suggestion (or a prefix of it) does not appear in it. But this only makes sense if copilot is correct more often than autocomplete.

**Comment by hsenag** [comment]
I think there are two main questions here:

- Do we ever want autocomplete and inline suggestions to show up simultaneously? The Copilot team has some experience from a while ago that it was bad, but that may be outdated.
- If they can show up simultaneously, then it's reasonably clear what to do if they don't conflict (discussed extensively above). The question is what to do if they do conflict. Should one win? I would say not and I also think the dummy entry would be a good way to resolve it.

**Comment by orph** [comment]
@hediet as @hsenag points out, I think there is some mixing questions here that I'd like to clarify Copilot's position on:

1) early user testing showed emphatically that mixing ghost text and autocomplete was visually overwhelming and ideologically confusing - people see them as different things
2) When we stopped trying to mix the two our complaints stopped
3) We have a handful of requests on the user forums that ask for showing autocomplete more often - not asking to integrate autocomplete with inline suggestions
4) at this time Copilot prefers to keep these features as separate as possible, with the guiding principle of “prefer inline suggestions but when the inline suggestion is wrong switch to autocomplete list”
5) we’re happy to revisit the early finding post November with a/b experiments and user interviews, but if the default setting is changed before we can finalize these finding we will have to override it
6) the algorithm I [proposed above](https://github.com/microsoft/vscode/issues/131940#issuecomment-931815784) is an incremental improvement on the principle in (4), and addresses the “show autocomplete more often” user request in (3) — we hope it will be considered in the short term to improve Copilot as we ramp up users

**Comment by hediet** [comment]
After discussing with @isidorn, we suggest the following:
* Keep the behavior of the last (1.60) release for now (in any case for 1.61 that gets released this week and currently equals the current insiders build, but also for future releases). This means the suggest widget is being suppressed when there is an inline suggestion.
* Wait for feedback of the intelli-code team and then discuss further steps and revisit this discussion.

**Comment by orph** [comment]
@hediet @isidorn  should I file another issue to get this comment implemented? https://github.com/microsoft/vscode/issues/131940#issuecomment-931815784 

We'd really like to have this as a UX improvement over the current situation to show the popup list more often when the inline suggestion isn't a good match.

**Comment by Christopher-Hayes** [comment]
That config setting helped me. In my case, I find TailwindCSS's Intellisense super useful and Copilot was suppressing it.

**Comment by dannyxnda** [comment]
When copilot is enable in the project, NO other suggestion work except for the inline suggestion created by Copilot.

**Comment by Christopher-Hayes** [comment]
I don't think the setting is fully documented, but it still seems to be used: https://github.com/microsoft/vscode/blob/92d25e35d9bf1a6b16f7d0758f25d48ace11e5b9/src/vs/editor/contrib/suggest/browser/suggestModel.ts#L112

**Comment by dannyxnda** [comment]
Yeah thanks, I see the implementation.

**Comment by jonahkagan** [comment]
Hey all, I wanted to share my thoughts as a user who's been struggling with this problem for a while now. I've mainly felt this pain when using VS Code with TypeScript IntelliSense and Copilot.

I've found that IntelliSense and Copilot inline suggestions serve very different purposes in my coding flow. IntelliSense tells me _what's possible_ while Copilot tells me _a best guess solution_. I use IntelliSense when I'm not totally sure what I want to do but I want to see what my options are. I use Copilot when I know what I want but I don't want to type it out. IntelliSense is like searching documentation intersected with the knowledge of the type system. Copilot is like pairing with someone who knows you so well they can finish your sentences for you.

Given that, my ideal behavior would be:
- Let Copilot keep making ghost text suggestions whenever it feels like it
- After typing a dot, show me IntelliSense and hide the Copilot ghost text

This would allow me to keep relying on Copilot to help speed up my typing but have a seamless way to indicate to VS Code when I'm interested in IntelliSense (typing a dot). (There are also other IntelliSense trigger characters where it would be useful to hide Copilot - e.g. typing an opening quote mark in a place where the type system knows a union of strings is being used.)

A few thoughts on other suggestions from this thread:
- Waiting til I type a character after the dot won't work well in cases where I need to browse all the available methods on an object (which happens quite often, e.g. when using libraries).
- I don't think it would bother me too much to show both Copilot ghost text and IntelliSense suggestions at the same time, but reading above that it was confusing for others seems like good user feedback that should be honored.
- I don't think Copilot suggestions should be shown in the IntelliSense suggestion pop-up, since they are distinctly different kinds of suggestions. I want to maintain my mental model of ghost text being moderately-reliable AI suggestions and the popup being extremely-reliable type system suggestions.

I'm sure I have some blind spots as I'm just one user with a specific use case, but hopefully this feedback can help as you keep working on this problem.

**Comment by frec-james** [comment]
I don't see this mentioned anywhere here but an acceptable workaround is to manually trigger IntelliSense with a keyboard shortcut: `Ctrl + space`.

**Comment by Shaun1** [comment]
Just stumbling upon this thread after trying out Copilot with Golang in VSCode for a few weeks now. I think @jonahkagan did a good job of outlining what an intuitive behavior should be, at least for my workflow. The Copilot suggestions are great most of the time but when I hit certain trigger points (such as a dot) I nearly always want to see what my valid available options are and select from them, rather than have a suggestion that may not be valid.

Maybe it's just me though. Just one more data point.

> I don't see this mentioned anywhere here but an acceptable workaround is to manually trigger IntelliSense with a keyboard shortcut: Ctrl + space.

This might work for some people, personally I've gotten used to IntelliSense suggestions coming up automatically at certain points without having to remember to hit another key.

**Comment by lcrownover** [comment]
IMO, the best user experience would be having copilot code as list items in the suggestions window. Combine that with the [suggestion-preview](https://code.visualstudio.com/updates/v1_57#_suggestion-preview) setting and you'd be able to mix and match LSP and Copilot suggestions.

Take a look at [this neovim plugin](https://github.com/zbirenbaum/copilot-cmp) to see examples of this exact behavior. It works very well!

**Comment by lukaszciastko** [comment]
I've been using Copilot for the past few weeks, and, this is, by far, the most annoying thing about it. **Could you at least make Copilot suggest valid things**?

For example, this is Copilot suggestion for the Flutter launch_url plugin:

<img width="360" alt="Screenshot 2023-04-21 at 12 32 20" src="https://user-images.githubusercontent.com/964414/233541309-dd133d09-6365-436a-ba80-64bc075ecea3.png">

However, the LaunchMode has no such option:

```
enum LaunchMode {
  /// Leaves the decision of how to launch the URL to the platform
  /// implementation.
  platformDefault,

  /// Loads the URL in an in-app web view (e.g., Safari View Controller).
  inAppWebView,

  /// Passes the URL to the OS to be handled by another application.
  externalApplication,

  /// Passes the URL to the OS to be handled by another non-browser application.
  externalNonBrowserApplication,
}
```

This makes no sense, and it happens over and over again, making using Copilot very frustrating and slowing down development.

**Comment by josias-r** [comment]
Just gonna drop this here for Copilot users:
```json
[
  {
    "key": "alt+right",
    "command": "editor.action.inlineSuggest.showNext",
    "when": "textInputFocus && inlineSuggestionVisible"
  },
  {
    "key": "alt+left",
    "command": "editor.action.inlineSuggest.showPrevious",
    "when": "inlineSuggestionVisible && textInputFocus"
  },
  {
    "key": "alt+down",
    "command": "editor.action.triggerSuggest",
    "when": "editorHasCompletionItemProvider && textInputFocus && !editorReadonly && !suggestWidgetVisible && inlineSuggestionVisible"
  }
]
```

These are the keybindings I use with vscode which really work great for me and solves this issue for me.

The important one is `alt+down`, this will show the suggestions (there is a default shortcut for that, but at least on a swiss keyboard it is HORRIBLE). And `alt+down` is just very intuitive for me -> open the dropd**down**.

The other two are to cycle through the copilot suggestions of you are not happy with the suggestion

**Comment by isidorn** [comment]
This should be much improved in the latest VS Code Stable. Can you all please try it out and let us know how it behaves for you?
To be sure you get the behaviour make sure to set `editor.inlineSuggest.suppressSuggestions: false` in your settings. 

Thanks all 👏 
I will go ahead and close this issue, but I can reopen if we see there are still improvements to be made.

**Comment by lcrownover** [comment]
Seems to behave much better!

**Comment by isidorn** [comment]
@lcrownover can you provide some details. What makes it better for you? Thanks!

**Comment by lcrownover** [comment]
@isidorn I successfully get the suggestions dropdown (LSP, etc) while the copilot ghost text is shown in the background. 
Behavior seems to match the preview in https://github.com/microsoft/vscode/issues/131940#issuecomment-914370898. 

I can hit `enter` or `tab` to accept the suggestion, then `tab` again to accept the copilot ghost text. (only needing one `tab` if no suggestion is shown).

**Comment by yangtzech** [comment]
> This should be much improved in the latest VS Code Stable. Can you all please try it out and let us know how it behaves for you? To be sure you get the behaviour make sure to set `editor.inlineSuggest.suppressSuggestions: false` in your settings.
> 
> Thanks all 👏 I will go ahead and close this issue, but I can reopen if we see there are still improvements to be made.

@isidorn There are still times when I expect IntelliCode to suggest methods when I hit a dot, but only CoPilot's suggestions appear.

I have attempted this on both VSCode 1.78.2 (system setup) and 1.79.0-insider (system setup), along with GitHub Copilot Nightly v1.86.118. All plugins are up-to-date.

**Comment by sebous** [comment]
> This should be much improved in the latest VS Code Stable. Can you all please try it out and let us know how it behaves for you? To be sure you get the behaviour make sure to set `editor.inlineSuggest.suppressSuggestions: false` in your settings.
> 
> Thanks all 👏 I will go ahead and close this issue, but I can reopen if we see there are still improvements to be made.

@isidorn Trying the new behaviour and it seems to be still broken, if copilot suggestion kicks in before I hit the `.`, LSP suggestions are not shown

![2023-06-01 13 38 05](https://github.com/microsoft/vscode/assets/14926669/bb8f5dfe-e584-437b-9821-c4d486260f5b)

Any idea what could be wrong?

env:
- vscode 1.78.2 (stable) - all extensions disabled except for GH 
- GH copilot nightly v1.87.131
- macOS 13.4
- all extensions disabled except for copilot
- all vscode settings commented out except for `editor.inlineSuggest.suppressSuggestions: false`

**Comment by hediet** [comment]
This should work for all extensions who don't set the suppressSuggestions flag in the returned inline completions result. Copilot sets it to `true`.

**Comment by sebous** [comment]
@hediet So is there a way to get it working? Or do I have to choose between copilot and language server suggestions showing automatically (without using the suggestions shortcut)?
Thanks a lot

**Comment by EntilZha** [comment]
What I'd really like to see are Copilot suggestions appearing in the intellisense dropdown, with configuration options on how many to show, the position in the list to use, etc (e.g., always insert the top copilot suggestion at the top, then N intellisense results, then Y copilot results, then the rest of intellisense). I think it would be worth re-opening this or creating a new issue to that end.

The real issue with the copilot UI is that it is fighting the already-mature intellisense dropdown rather than having suggestions appear there.

**Comment by isidorn** [comment]
Just to let you know that we have pushed a fix so that now suggest widget is NOT suppressed by default.
So this should work better with VS Code Insiders from tomorrow. And this will be included in next stable (start of September). Try it out and let us know what you think. Thank you!
